migrator.py

```diff
@@ -3,39 +3,153 @@
+    ini = configparser.ConfigParser(strict=False)
+    ini_file = input_dir /  Path(ini_file_name)
+    ini.read(ini_file, encoding='utf8')
+    try:
+        for contact_id, value in ini['Contacts'].items():
+            person_name = ".NoName-"+value.split(',')[1] # this is actually a hash
+            if not contact_id in contact_id_2_nameset:
+                logging.info(f"Learned name (old picasa contact ID) for {contact_id}='{person_name}' from {ini_file}")
+                contact_id_2_nameset[contact_id]=set([person_name])
+            # Note: the difference between learn_contact_ids and this version is we don't add the name to the set
+    except KeyError:
+        # likely no Contacts2 section
+        pass
+    except Exception as e:
+        logging.warning("Exception while learning contact ids")
+        logging.exception(str(e))
+
+def migrate_directories_under(input_root_dir: Path, 
+                              db: DigikamDb, 
+                              dry_run: bool,
+                              contacts_file: Path):
     """Traverses directory tree to find directories to migrate."""
+
+    # Build a contact_id to name(s) dictionary
+    global_names: GlobalNames = {}
+
+    if contacts_file is not None:
+        # contacts.xml seems more authoratative than *.ini files
+        # Use this whenever possible.
+        # Can be found at %LocalAppData%\Google\Picasa2\contacts\contacts.xml
+        # Also can be obtained from performing a backup operation from 
+        # within Picasa.  It will be in the backup location as 
+        # $Application Data\Google\Picasa2\contacts\backup.xml
+        logging.info(f'Reading contacts from {contacts_file}')
+        tree = ET.parse(contacts_file)
+        for contact in tree.getroot():
+            logging.info(f'{contact.attrib}')
+            global_names[contact.attrib['id']] = contact.attrib['name']
+
+    else:
+        # Alternate way using *.ini files
+        # Sometimes names will be missing or conflicting if found 
+        # different in multiple directories.  Conflicting names
+        # will be merged by this script by concatenating them with " | " 
+        # separators in sorted order.  Missing names will have an
+        # auto generated value of  
+        contact_id_2_nameset: Dict[str, set] = {}
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+                learn_contact_ids(input_dir,ini_file,contact_id_2_nameset)
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+                learn_old_contact_ids(input_dir,ini_file,contact_id_2_nameset)
+
+        logging.info("contact_id_2_nameset")
+        logging.info("------------------------------------------------------------")
+        try:
+            for contact_id, names in contact_id_2_nameset.items():
```

Comment:
There's a lot going on here with indices and enumeration that makes it hard to follow. Please try to express this in functional rather than procedural style.

Is it just this?:

```
global_names = {k: "|".join(sorted(v)) for k,v in contact_id_2_nameset.items()}
```